File: interface/src/models/FundingSchedule.ts
import type BankAccount from '@monetr/interface/models/BankAccount';
import { ID, idPrefix } from '@monetr/interface/models/ID';
import type { WithJsonValues } from '@monetr/interface/util/json';
import parseDate from '@monetr/interface/util/parseDate';

export default class FundingSchedule {
  readonly [idPrefix] = 'fund';

  readonly fundingScheduleId: ID<FundingSchedule>;
  readonly bankAccountId: ID<BankAccount>;
  name: string;
  description: string | null;
  ruleset: string;
  readonly lastRecurrence: Date | null;
  nextRecurrence: Date;
  readonly nextRecurrenceOriginal: Date;
  excludeWeekends: boolean;
  autoCreateTransaction: boolean;
  estimatedDeposit: number | null;

  constructor(data: WithJsonValues<FundingSchedule>) {
    this.fundingScheduleId = ID.from(data.fundingScheduleId);
    this.bankAccountId = ID.from(data.bankAccountId);
    this.name = data.name;
    this.description = data.description;
    this.ruleset = data.ruleset;
    this.lastRecurrence = data.lastRecurrence ? parseDate(data.lastRecurrence) : null;
    this.nextRecurrence = parseDate(data.nextRecurrence);
    this.nextRecurrenceOriginal = parseDate(data.nextRecurrenceOriginal);
    this.excludeWeekends = data.excludeWeekends;
    this.autoCreateTransaction = data.autoCreateTransaction;
    this.estimatedDeposit = data.estimatedDeposit;
  }
}

File: server/schema/funding_schedule.go
package schema

import (
	"github.com/Oudwins/zog"
	"github.com/monetr/monetr/server/models"
)

var (
	CreateFundingSchedule = zog.Struct(zog.Shape{
		"name":                  Name(),
		"description":           Description(),
		"ruleset":               zog.Ptr(RRule()).NotNil(),
		"excludeWeekends":       zog.Bool().Default(false).Required(),
		"estimatedDeposit":      zog.Ptr(zog.Int64().GT(0).Optional()),
		"nextRecurrence":        Date().Required(),
		"autoCreateTransaction": zog.Bool().Default(false).Required(),
	})

	PatchFundingSchedule = zog.Struct(zog.Shape{
		"name":                  Name().Optional(),
		"description":           Description().Optional(),
		"ruleset":               zog.Ptr(RRule()),
		"excludeWeekends":       zog.Bool().Optional(),
		"estimatedDeposit":      zog.Ptr(zog.Int64().GT(0).Optional()),
		"nextRecurrence":        Date().Optional(),
		"autoCreateTransaction": zog.Bool().Optional(),
	})
)

func isValidFundingSchedule(val any, ctx zog.Ctx) bool {
	switch val := val.(type) {
	case *models.FundingSchedule:
		if val.AutoCreateTransaction &&
			(val.EstimatedDeposit == nil || *val.EstimatedDeposit <= 0) {
			ctx.AddIssue(ctx.Issue().
				SetCode("invalid_autoCreateTransaction").
				SetPath([]string{"autoCreateTransaction"}).
				SetMessage("auto create transaction requires that an estimated deposit is specified").
				SetParams(map[string]any{
					"estimatedDeposit":      val.EstimatedDeposit,
					"autoCreateTransaction": val.AutoCreateTransaction,
				}),
			)
		}
	}

	return true
}
